### PR TITLE
Form - render rootLayoutManager

### PR DESCRIPTION
**Pull request by novsstation**

### CODE REVIEW

File: js/renovation/ui/form/screen_utils.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { ScreenSize } from './types';
+
+export const defaultScreenFactorFunc = (width: number): ScreenSize => {
```
Comment (IgnatovDan) [inline js/renovation/ui/form/screen_utils.ts]: defaultScreenFactorFunc -> convertToScreenSizeQualifier ?

File: js/renovation/ui/form/types.d.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+export type ScreenSize = 'xs' | 'sm' | 'md' | 'lg';
```
Comment (IgnatovDan) [inline js/renovation/ui/form/types.d.ts]: ScreenSize -> ScreenSizeQualifier ?
https://js.devexpress.com/Documentation/Guide/UI_Components/ResponsiveBox/Size_Qualifiers/

File: js/renovation/ui/form/form.tsx [hunk 3]
```diff
@@ -1,24 +1,31 @@
...
 
 import { FormProps } from './form_props';
 
 import { combineClasses } from '../../utils/combine_classes';
 import { Widget } from '../common/widget';
+import { FormLayoutManager } from './form_layout_manager';
 import { Scrollable } from '../scroll_view/scrollable';
 
+const renderLayoutManager = (isRoot: boolean): JSX.Element => (
+  <FormLayoutManager isRoot={isRoot} />
```
Comment (IgnatovDan) [inline js/renovation/ui/form/form.tsx]: FormLayoutManager -> LayoutManager ?

const instance = that._createComponent('dxLayoutManager', config); - we are moving this line

File: js/renovation/ui/form/form.tsx [hunk 4]
```diff
@@ -1,24 +1,31 @@
...
 } from '@devextreme-generator/declarations';
 
 import { FormProps } from './form_props';
 
 import { combineClasses } from '../../utils/combine_classes';
 import { Widget } from '../common/widget';
+import { FormLayoutManager } from './form_layout_manager';
 import { Scrollable } from '../scroll_view/scrollable';
 
+const renderLayoutManager = (isRoot: boolean): JSX.Element => (
```
Comment (IgnatovDan) [inline js/renovation/ui/form/form.tsx]: isRoot is not used
Let's remove it or use it in this PR

File: js/renovation/ui/form/screen_utils.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { ScreenSize } from './types';
```
Comment (IgnatovDan) [inline js/renovation/ui/form/screen_utils.ts]: Are we need it in this PR?

File: js/renovation/ui/form/types.d.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
```
Comment (IgnatovDan) [inline js/renovation/ui/form/types.d.ts]: Are we need it in this PR?

File: js/renovation/ui/form/__tests__/form.test.tsx [hunk 7]
```diff
@@ -37,15 +38,24 @@ describe('Form > Markup', () => {
...
         direction: 'both',
         useKeyboard: false,
         useNative: !!useNativeScrolling,
         useSimulatedScrollbar: !useNativeScrolling,
-      });
+      }));
+    });
+
+    it('root layoutManager is rendered', () => {
+      const props = { scrollingEnabled: true, useNativeScrolling } as FormProps;
```
Comment (IgnatovDan) [inline js/renovation/ui/form/__tests__/form.test.tsx]: scrollingEnabled: false ?

File: js/renovation/ui/form/__tests__/form.test.tsx [hunk 8]
```diff
@@ -37,15 +38,24 @@ describe('Form > Markup', () => {
...
+      const form = mount<Form>(<Form {...props} />);
+
+      const layoutManager = form.find(FormLayoutManager);
+      expect(layoutManager.exists()).toEqual(true);
+      expect(layoutManager.props().isRoot).toEqual(true);
```
Comment (IgnatovDan) [inline js/renovation/ui/form/__tests__/form.test.tsx]: isRoot is not used in code, are we need it in this PR?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/renovation/ui/form/form.tsx [hunk 9]
```diff
@@ -31,15 +38,19 @@ export const viewFunction = (viewModel: Form): JSX.Element => {
...
     )
     : (
       <Widget
         aria={aria}
         classes={cssClasses}
         // eslint-disable-next-line react/jsx-props-no-spreading
         {...restAttributes}
-      />
+      >
+        {rootLayoutManager}
```
Comment (IgnatovDan) [inline js/renovation/ui/form/form.tsx:48]: Whether we should replace Widget with rootLayoutManager ?

In old code it was:
    _getContent: function() {
        return this.option('scrollingEnabled') ? this._scrollable.$content() : this.$element();
    },

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/renovation/ui/form/form.tsx [hunk 10]
```diff
@@ -31,15 +38,19 @@ export const viewFunction = (viewModel: Form): JSX.Element => {
         bounceEnabled={false}
         // eslint-disable-next-line react/jsx-props-no-spreading
         // TODO: {...restAttributes} waits for the https://trello.com/c/er8aTcsZ/2711-renovation-react-some-events-from-restattributes-may-have-the-same-type-as-react-exists-events
-      />
+      >
```
Comment (IgnatovDan) [inline js/renovation/ui/form/form.tsx:37]: Whether we should apply these aria/classes in this section?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/renovation/ui/form/form.tsx [hunk 11]
```diff
@@ -1,24 +1,31 @@
...
 
 import { FormProps } from './form_props';
 
 import { combineClasses } from '../../utils/combine_classes';
 import { Widget } from '../common/widget';
+import { LayoutManager } from './form_layout_manager';
 import { Scrollable } from '../scroll_view/scrollable';
 
+const renderLayoutManager = (): JSX.Element => (
+  <LayoutManager />
```
Comment (IgnatovDan) [inline js/renovation/ui/form/form.tsx]: Let's expand it into an inline code.

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/renovation/ui/form/form.tsx (hunk 10 above)
Comment (IgnatovDan) [reply js/renovation/ui/form/form.tsx:37]: Already applied:
      <Scrollable
        aria={aria}

File: js/renovation/ui/form/form.tsx (hunk 9 above)
Comment (novsstation) [reply js/renovation/ui/form/form.tsx:48]: No, no. When scrolling is disabled, then layout manager is wrapped by the div with aria attributes and "dx-widget form" css classes
image


In the old code, when scrolling is enabled, then scrollable is applied to the main div. So main div is combination of the 2 widgets: form and scrollable - https://github.com/DevExpress/DevExtreme/blob/37ab2c57d466b499a460557cd1bed95fc9ef29d9/js/ui/form/ui.form.js#L357

File: js/renovation/ui/form/form.tsx (hunk 11 above)
Comment (novsstation) [reply js/renovation/ui/form/form.tsx]: fixed

File: js/renovation/ui/form/__tests__/form.test.tsx (hunk 8 above)
Comment (novsstation) [reply js/renovation/ui/form/__tests__/form.test.tsx]: fixed

File: js/renovation/ui/form/__tests__/form.test.tsx (hunk 7 above)
Comment (novsstation) [reply js/renovation/ui/form/__tests__/form.test.tsx]: fixed

File: js/renovation/ui/form/types.d.ts (hunk 6 above)
Comment (novsstation) [reply js/renovation/ui/form/types.d.ts]: fixed

File: js/renovation/ui/form/screen_utils.ts (hunk 5 above)
Comment (novsstation) [reply js/renovation/ui/form/screen_utils.ts]: fixed

File: js/renovation/ui/form/form.tsx (hunk 4 above)
Comment (novsstation) [reply js/renovation/ui/form/form.tsx]: fixed

File: js/renovation/ui/form/types.d.ts (hunk 2 above)
Comment (novsstation) [reply js/renovation/ui/form/types.d.ts]: I removed this whole file

File: js/renovation/ui/form/screen_utils.ts (hunk 1 above)
Comment (novsstation) [reply js/renovation/ui/form/screen_utils.ts]: I removed this whole file

**Review by IgnatovDan — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Pull request merged**